.tools/validate.go

```diff
@@ -0,0 +1,196 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+var (
+	flCommitRange = flag.String("range", "", "use this commit range instead")
+
+	validDCO = regexp.MustCompile(`^Signed-off-by: ([^<]+) <([^<>@]+@[^<>]+)>$`)
+)
+
+var DefaultRules = []ValidateRule{
+	func(c CommitEntry) error {
+		hasValid := false
+		for _, line := range strings.Split(c.Body, "\n") {
+			if validDCO.MatchString(line) {
+				hasValid = true
+			}
+		}
+		if !hasValid {
+			return fmt.Errorf("%q does not have a valid DCO", c.Commit)
+		}
+		return nil
+	},
+	// TODO add something for the cleanliness of the c.Subject
+}
+
+func main() {
+	flag.Parse()
+
+	/*
+		if len(os.Getenv("TRAVIS")) > 0 {
+			fmt.Println(os.Getenv("TRAVIS_COMMIT_RANGE"))
+			if len(os.Getenv("TRAVIS_PULL_REQUEST")) > 0 {
+				fmt.Println(os.Getenv("TRAVIS_BRANCH"))
+			}
+		}
+	*/
+
+	invalids := []InvalidCommit{}
+	if *flCommitRange != "" {
+		c, err := GitCommits(*flCommitRange)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		for _, commit := range c {
+			fmt.Printf(" * %s %s ... ", commit.AbbreviatedCommit, commit.Subject)
+			if err := ValidateCommit(commit, DefaultRules); err != nil {
+				invalids = append(invalids, InvalidCommit{commit, err})
+				fmt.Printf("FAIL: %s\n", err)
+			} else {
+				fmt.Println("PASS")
+			}
+		}
+	} else if os.Getenv("TRAVIS_COMMIT_RANGE") != "" {
+		c, err := GitCommits(os.Getenv("TRAVIS_COMMIT_RANGE"))
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		for _, commit := range c {
+			fmt.Printf(" * %s %s ... ", commit.AbbreviatedCommit, commit.Subject)
+			if err := ValidateCommit(commit, DefaultRules); err != nil {
+				invalids = append(invalids, InvalidCommit{commit, err})
+				fmt.Printf("FAIL: %s\n", err)
+			} else {
+				fmt.Println("PASS")
+			}
+		}
+	}
+	if len(invalids) > 0 {
+		fmt.Printf("%d issues to fix\n", len(invalids))
+		os.Exit(1)
+	}
+}
+
+type ValidateRule func(CommitEntry) error
+
+func ValidateCommit(c CommitEntry, rules []ValidateRule) error {
+	for _, r := range rules {
+		if err := r(c); err != nil {
+			return err
```

Comment:
Instead of failing after the first invalid rule, maybe always iterate over all rules and return an array of failures?  Otherwise there will be noise in PRs as folks fix the initial error, repush, get notified of the previously-masked error, fix it, repush, ….